Feature/secret (#3)
figment-networks/datahub-learn: network-documentation/secret/tutorials/4.-submit-your-first-transaction.md

```javascript
const { logs, transactionHash } = await client.postTx(signedTx);
```

### Query transaction
Under the comment `// 5. Query transaction` add the following code snippet below:

```javascript
const query = {id: transactionHash}
const tx = await client.searchTx(query)
console.log('Transaction: ', tx);
```

Let's run the code:

```text
node send.js
```

If everything works \(you have enough tokens, recipient is valid, etc\) you should see an output similar to:

```javascript
Transaction:  [
  {
    height: 400136,
    hash: '7C3C689EA4CEAD0C89BE039063909307C5C9048F76DF05E387E0368BD1C4435C',
    code: 0,
    rawLog: '[{"msg_index":0,"log":"","events":[{"type":"message","attributes":[{"key":"action","value":"send"},{"key":"sender","value":"secret1cdycaskx8g9gh9zpa5g8ah04ql0lzkrsxmcnfq"},{"key":"module","value":"bank"}]},{"type":"transfer","attributes":[{"key":"recipient","value":"secret1cdycaskx8g9gh9zpa5g8ah04ql0lzkrsxmcnfq"},{"key":"amount","value":"1000000uscrt"}]}]}]',
    logs: [ [Object] ],
    tx: { type: 'cosmos-sdk/StdTx', value: [Object] },
    timestamp: '2020-12-05T07:42:52Z'
  }

> ]

Reviewer: missing ending brackets for ```javascript

2. [Setting up your wallet](https://learn.figment.io/network-documentation/secret/tutorials/2.-setting-up-your-wallet.md)
3. [Querying the Secret Network](https://learn.figment.io/network-documentation/secret/tutorials/3.-querying-the-secret-blockchain)

We will be building on top of the Node.js application created in these tutorials.

In the previous tutorials, we created our account and we used the Testnet faucet to request test tokens. Now that we have some tokens to spend, let's send some of them to another account.

Ok, let’s move on to writing our first transaction!

### Simple Transfers

Start by creating a new file `transfer.js` in the project directory and add the code below:

```
const {
    EnigmaUtils, SigningCosmWasmClient, Secp256k1Pen, pubkeyToAddress, encodeSecp256k1Pubkey
} = require("secretjs");

require('dotenv').config();

const main = async () => {
    
    const mnemonic = process.env.MNEMONIC;
    const httpUrl = process.env.SECRET_NODE_URL;
    const signingPen = await Secp256k1Pen.fromMnemonic(mnemonic);
    const pubkey = encodeSecp256k1Pubkey(signingPen.pubkey);
    const accAddress = pubkeyToAddress(pubkey, 'secret');

>     const txEncryptionSeed = EnigmaUtils.GenerateNewSeed();

Reviewer: ```suggestion
```

removing because it's also declared in the next code snippet

```javascript
const query = {id: sent.transactionHash};
const tx = await client.searchTx(query);
console.log('Transaction: ', tx);
```

Let's run the code:

```text
node transfer.js
```

If everything works \(you have enough tokens, recipient is valid, etc\) you should see an output similar to:

```javascript
Transaction:  [
  {
    height: 504795,
    hash: '8ED3C35EA83EF2A016AF900172563E0D7ECFA2AC00C616714EF3C68FC9AE9526',
    code: 0,
    rawLog: '[{"msg_index":0,"log":"","events":[{"type":"message","attributes":[{"key":"action","value":"send"},{"key":"sender","value":"secret1cdycaskx8g9gh9zpa5g8ah04ql0lzkrsxmcnfq"},{"key":"module","value":"bank"}]},{"type":"transfer","attributes":[{"key":"recipient","value":"secret1cdycaskx8g9gh9zpa5g8ah04ql0lzkrsxmcnfq"},{"key":"amount","value":"1234uscrt"}]}]}]',
    logs: [ [Object] ],
    tx: { type: 'cosmos-sdk/StdTx', value: [Object] },
    timestamp: '2020-12-12T06:43:19Z'
  }
]
```
----------------------------------------------------------------
Copy the transaction's `hash` from the output and replace `<TRANSACTION HASH>` in the link below, to see the transaction in Secret Explorer!

> https://explorer.secrettestnet.io/transactions/<TRANSACTION HASH>

Reviewer: `<>` is being interpreted as html and not rendering in the markdown. Maybe we should just show the link as a string, ie`https://explorer.secrettestnet.io/transactions/<TRANSACTION HASH>`

Author: Please see if that looks better

Reviewer: @Y-Folla do you think the links looks ok like this?

https://github.com/figment-networks/datahub-learn/blob/cc415875325a5ba724a48a240961d7f27dd23a2f/network-documentation/secret/tutorials/4.-submit-your-first-transaction.md#copy-the-transactions-hash-from-the-output-and-replace-transaction-hash-in-the-link-below-to-see-the-transaction-in-secret-explorerhttpsexplorersecrettestnetiotransactionstransaction-hash

Reviewer 2: yeah that works

Author: Got the last one too

    // 4. Broadcast TX

    // 5. Query TX by hash/ID
}

main().then(resp => {
    console.log(resp);
}).catch(err => {
    console.log(err);
})

```
By now this should look familiar to you. Starting from the top you can see that we import secretjs, dotenv, and then create a new instance using `CosmWasmClient`. This establishes the connection to our Secret Network node. 
We then get the network and account information.

### Define TX message
In the `send.js` file under the comment `// 1. Define TX message` add the following code snippet below:

```javascript
const sendMsg = {
    type: "cosmos-sdk/MsgSend",
    value: {
        from_address: accAddress,
        to_address: accAddress,
        amount: [
            {
                denom: "uscrt",
                amount: "1000000",
            },
        ],
    },
};
```

### Define fees
Under the comment `// 2. Define fees` add the following code snippet below:

```javascript
const fee = {
    amount: [
        {
            amount: "50000",
            denom: "uscrt",
        },
    ],
    gas: "100000",
};
```

### Sign transaction
Under the comment `// 3. Sign transaction` add the following code snippet below:

```javascript

> const signBytes = makeSignBytes([sendMsg], fee, chainId, memo, accountNumber, sequence);

Reviewer: accountNumber was undefined, missing line:
```suggestion
const { accountNumber, sequence } = await client.getNonce(accAddress);
const signBytes = makeSignBytes([sendMsg], fee, chainId, memo, accountNumber, sequence);
```

```javascript
Transaction:  [
  {
    height: 400136,
    hash: '7C3C689EA4CEAD0C89BE039063909307C5C9048F76DF05E387E0368BD1C4435C',
    code: 0,
    rawLog: '[{"msg_index":0,"log":"","events":[{"type":"message","attributes":[{"key":"action","value":"send"},{"key":"sender","value":"secret1cdycaskx8g9gh9zpa5g8ah04ql0lzkrsxmcnfq"},{"key":"module","value":"bank"}]},{"type":"transfer","attributes":[{"key":"recipient","value":"secret1cdycaskx8g9gh9zpa5g8ah04ql0lzkrsxmcnfq"},{"key":"amount","value":"1000000uscrt"}]}]}]',
    logs: [ [Object] ],
    tx: { type: 'cosmos-sdk/StdTx', value: [Object] },
    timestamp: '2020-12-05T07:42:52Z'
  }
]
```
----------------------------------------------------------------
Copy the transaction's `hash` from the output and replace `<TRANSACTION HASH>` in the link below.
OPEN LINK BELOW to see the transaction in Secret Explorer!
https://explorer.secrettestnet.io/transactions/<TRANSACTION HASH>
----------------------------------------------------------------

## Conclusion

Great job! Now you not only know how to read from the Secret Network, but also how to interact with it by crafting transactions from scratch with SecretJS and DataHub.

> The complete code can be found on [**Github**](https://github.com/figment-networks/tutorials/blob/main/secret/4_send_transaction/main.js)**.** 

Reviewer: GitHub link doesn't exist, probably want this
```suggestion
The complete code can be found on [**Github**](https://github.com/figment-networks/tutorials/blob/main/secret/4_transactions)**.** 
```
